tests/epyccel/test_epyccel_lists.py

```diff
@@ -117,3 +117,79 @@ def f():
 
     epyc_f = epyccel(f, language=language)
     assert f() == epyc_f()
+
+def test_insert_basic(language):
+    def f():
+        a = [1, 2, 3]
+        a.insert(4, 4)
+        return a
+
+    epyc_f = epyccel(f, language=language)
+    assert f() == epyc_f()
+
+def test_insert_multiple(language):
+    def f():
+        a = [1, 2, 3]
+        a.insert(4, 4)
+        a.insert(2, 5)
+        a.insert(1, 6)
+        return a
+
+    epyc_f = epyccel(f, language=language)
+    assert f() == epyc_f()
+
+def test_insert_list(language):
+    def f():
+        a = [[1, 2, 3]]
+        a.insert(1, [4, 5, 6])
+        return a
+
+    epyc_f = epyccel(f, language=language)
+    assert f() == epyc_f()
+
+def test_insert_range(language):
+    def f():
+        a = [1, 2, 3]
+        for i in range(4, 1000):
+            a.insert(i - 1 ,i)
+        return a
+
+    epyc_f = epyccel(f, language=language)
+    assert f() == epyc_f()
+
+def test_insert_range_list(language):
+    def f():
+        a = [[1, 2, 3]]
+        for i in range(4, 1000):
+            a.insert(i, [i, i + 1])
+        return a
+
+    epyc_f = epyccel(f, language=language)
+    assert f() == epyc_f()
+
+def test_insert_range_tuple(language):
+    def f():
+        a = [[1, 2, 3]]
+        for i in range(4, 1000):
+            a.insert(i, (i, i + 1))
+        return a
+
+    epyc_f = epyccel(f, language=language)
+    assert f() == epyc_f()
+
+def test_insert_user_defined_objects(language):
+    def f():
+        class A:
+            pass
+        a = A()
+        b = A()
+        c = A()
+        d = A()
+        e = A()
+        lst = [a, b, c]
+        lst.insert(0, d)
+        lst.insert(1, e)
+        return a
```

Comment:
This is a really nice idea but it seems that it is crashing your tests. I had a quick look and Pyccel doesn't currently handle classes inside functions (but unfortunately it doesn't raise a neat error about this). If you want to write a test like this then you will need to create a test module:
```python
class A:
    pass
def f():
    a = A()
    b = A()
    c = A()
    d = A()
    e = A()
    lst = [a, b, c]
    lst.insert(0, d)
    lst.insert(1, e)
    return a
```
and translate the whole module

Reply:
The problem is that `mod.A` and `modnew.A` are not the same type even if they look like it. For this you can check the length of the list and you can check the properties of the class to make sure the list is correctly ordered. For example:
```python
class A:
    def __init__(self, x : int):
        self.x = x

def f():
    a = A(1)
    b = A(2)
    c = A(3)
    d = A(4)
    e = A(5)
    lst = [a, b, c]
    lst.insert(0, d)
    lst.insert(1, e)
    return a
```

```
def test_insert_user_defined_objects(language):
    import modules.list_user_defined_objs as mod

    modnew = epyccel(mod, language=language)
    python_list = mod.fn()
    accelerated_list = modnew.fn()
    assert len(python_list) == len(accelerated_list)
    for python_elem, accelerated_elem in zip(python_list, accelerated_list):
         assert python_elem.x == accelerated_elem.x
```

Reply:
tried this:
```python
class A:
    def __eq__(self, other):
        return isinstance(other, A)

def fn():
    a = A()
    b = A()
    c = A()
    d = A()
    e = A()
    lst = [a, b, c]
    lst.insert(0, d)
    lst.insert(1, e)
    return lst
```
but got a semantic error, i don't really know the extent of the classes support in Pyccel

Reply:
Classes support:
- variables
- basic functions

They don't support:
- magic methods
- decorators

I'm not sure we have support for `isinstance` either actually

Reply:
fixed in 34d82c90fff7be918ee42328813295c2bee65ca9